src/fomo.jl

```diff
@@ -0,0 +1,399 @@
+  )
+
+  callback(nlp, solver, stats)
+
+  done = stats.status != :unknown
+
+  d .= ∇fk
+  norm_d = norm_∇fk
+  βmax = T(0)
+  ρk = T(0)
+  avgβmax = T(0)
+  siter = 0
+  oneT = T(1)
+  mdot∇f = T(0) # dot(momentum,∇fk)
+  while !done
+    λk = step_mult(solver.α,norm_d,step_backend)
+    c .= x .- λk .* d
+    step_underflow = x == c # step addition underfow on every dimensions, should happen before solver.α == 0
+    ΔTk = ((oneT - βmax) * norm_∇fk^2 + βmax * mdot∇f) * λk # = dot(d,∇fk) * λk with momentum, ‖∇fk‖²λk without momentum
+    fck = obj(nlp, c)
+    if fck == -Inf
+      set_status!(stats, :unbounded)
+      break
+    end
+    ρk = (stats.objective - fck) / ΔTk
+    # Update regularization parameters
+    if ρk >= η2
+      solver.α = min(αmax, γ2 * solver.α)
+    elseif ρk < η1
+      solver.α = solver.α * γ1
+      if use_momentum
+        βmax *= γ3
+        d .= ∇fk .* (oneT - βmax) .+ momentum .* βmax
+      end
+    end
+
+    # Acceptance of the new candidate
+    if ρk >= η1
+      x .= c
+      if use_momentum
+        momentum .= ∇fk .* (oneT - β) .+ momentum .* β
+        mdot∇f = dot(momentum,∇fk)
+      end
+      set_objective!(stats, fck)
+      grad!(nlp, x, ∇fk)
+      norm_∇fk = norm(∇fk)
+      if use_momentum
+        p .= momentum .- ∇fk
+        βmax = find_beta(p , mdot∇f, norm_∇fk, β, θ1, θ2)
+        d .= ∇fk .* (oneT - βmax) .+ momentum .* βmax
+        norm_d = norm(d)
+      end
+      if use_momentum
+        avgβmax += βmax
+        siter += 1
+      end
+    end
+
+    set_iter!(stats, stats.iter + 1)
+    set_time!(stats, time() - start_time)
+    set_dual_residual!(stats, norm_∇fk)
+    optimal = norm_∇fk ≤ ϵ
+
+    if verbose > 0 && mod(stats.iter, verbose) == 0
+      @info infoline
+      if !use_momentum
+        infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk 1/solver.α ρk
+      else
+        infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk solver.α ρk βmax
+      end
+    end
+
+    set_status!(
+      stats,
+      get_status(
+        nlp,
+        elapsed_time = stats.elapsed_time,
+        optimal = optimal,
+        max_eval = max_eval,
+        iter = stats.iter,
+        max_iter = max_iter,
+        max_time = max_time,
+      ),
+    )
+    
+    callback(nlp, solver, stats)
+
+    step_underflow  && set_status!(stats,:small_step)
+    solver.α == 0         && set_status!(stats,:exception) # :small_nlstep exception should happen before
+
+    done = stats.status != :unknown
+  end
+  if use_momentum
+    avgβmax /= siter
+    stats.solver_specific[:avgβmax] = avgβmax
+  end
+  set_solution!(stats, x)
+  return stats
+end
+
+"""
+find_beta(m, md∇f, norm_∇f, β, θ1, θ2)
+
+Compute βmax which saturates the contibution of the momentum term to the gradient.
+`βmax` is computed such that the two gradient-related conditions are ensured: 
+1. [(1-βmax) * ∇f(xk) + βmax * dot(m,∇f(xk))] ≥ θ1 * ‖∇f(xk)‖²
```

Comment:
This condition mixes vectors and scalars.

Reply:
Also, in the docstring, you can just write `∇f(xk)ᵀm`, which does not require an explanation. In fact, you may be able to just use that as variable name too.